### PR TITLE
[mqtt] Added option to define multiple listeners and use TLS

### PR DESCRIPTION
**Pull request by vipeller**
- multiple listeners can be spun up (e.g. on different ports) instead of a single one
- a listener (represented by "transports") can play either TLS or pure TCP

### CODE REVIEW

File: mqtt/mqtt-broker/src/transport.rs [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+use tracing::{debug, error, warn};
+
+use crate::{Error, ErrorKind};
+
+pub enum TransportBuilder<A> {
+    Tcp(A),
+    Tls(A, Identity),
+}
+
+impl<A> TransportBuilder<A>
```
[the remark this answers is not in the document]
Comment (vipeller) [reply mqtt/mqtt-broker/src/transport.rs]: ok, come back later to improve

File: mqtt/mqtt-broker/src/server.rs [hunk 2]
```diff
@@ -52,36 +63,91 @@ impl Server {
...
                     debug!("sending Shutdown message to broker");
+
+                    if let Err(e) = &result {
+                        error!(message = "an error occurred in the accept loop", error=%e);
+                    }
+
+                    debug!("sending stop signal for the rest of protocol heads");
+                    incoming_tasks_tx.remove(index);
+                    for itx in incoming_tasks_tx {
+                        itx.send(()).unwrap();
```
Comment (myagley) [inline mqtt/mqtt-broker/src/server.rs]: same comment about `unwrap`

File: mqtt/mqttd/src/main.rs [hunk 3]
```diff
@@ -19,10 +20,22 @@ async fn main() -> Result<(), Error> {
         .finish();
     let _ = tracing::subscriber::set_global_default(subscriber);
 
-    let addr = env::args()
+    let addr_tcp = env::args()
```
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqttd/src/main.rs]: It will be a config file to setup the broker. All it will accept is a path to a config file. Do we still need to use `clap`?

File: mqtt/mqtt-broker/src/server.rs [hunk 4]
```diff
@@ -28,22 +26,35 @@ impl Server {
...
-    pub async fn serve<A, F>(self, addr: A, shutdown_signal: F) -> Result<BrokerState, Error>
+    pub async fn serve<A, F>(
+        self,
+        transports: Vec<TransportBuilder<A>>,
+        shutdown_signal: F,
+    ) -> Result<BrokerState, Error>
     where
-        A: ToSocketAddrs + Display,
+        A: ToSocketAddrs,
         F: Future<Output = ()> + Unpin,
```
[the remark this answers is not in the document]
Comment (myagley) [reply mqtt/mqtt-broker/src/server.rs]: The other option is a builder pattern. Something like:

```rust
struct Server {
    broker: Broker
    transports: Vec<TransportBuilder>,
}

Server::new(broker)
    .transport(TransportBuilder::Tcp(addr))
    .transport(TransportBuilder::Tls(...))
    .serve(shutdown);
```

or

```rust
struct Server {
    broker: Broker
    transports: Vec<TransportBuilder>,
}

Server::new(broker)
    .tcp(addr)
    .tls(addr, identity)
    .serve(shutdown);
```

File: mqtt/mqttd/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (myagley) [reply mqtt/mqttd/src/main.rs]: I probably still would and put the config under a `-f` flag.

File: mqtt/mqtt-broker/src/server.rs [hunk 5]
```diff
@@ -28,22 +26,35 @@ impl Server {
...
+        let broker_task = tokio::spawn(broker.run());
 
-        let (itx, irx) = oneshot::channel::<()>();
+        let mut incoming_tasks = Vec::new();
+        let mut incoming_tasks_tx = Vec::new();
+        for transport in transports {
+            let (itx, irx) = oneshot::channel::<()>();
+            incoming_tasks_tx.push(itx);
+
+            let incoming_task = incoming_task(transport, handle.clone(), irx.map(drop));
```
[the remark this answers is not in the document]
Comment (myagley) [reply mqtt/mqtt-broker/src/server.rs]: ah never mind. shutdown signal is a `Future<()>`

File: mqtt/mqttd/src/main.rs [hunk 6]
```diff
@@ -70,6 +87,16 @@ async fn main() -> Result<(), Error> {
     Ok(())
 }
 
+async fn load_identity(path: String) -> Result<Identity, Error> {
+    let cert_buffer = fs::read(&path).await.context(ErrorKind::LoadIdentity)?;
```
[the remark this answers is not in the document]
Comment (vipeller) [reply mqtt/mqttd/src/main.rs]: but then what is rule of thumb deciding what should be async? I figured that if the caller is already async, and this one does a "slow" operation, that is how it became async, too. (something like async 'spreads' the async-ness downwards, it that makes sense)
I understand that it doesn't have any performance benefits in this case, it is more like a "don't block an async method" thing. Is there a better way to think about this?

File: mqtt/mqttd/Cargo.toml [hunk 7]
```diff
@@ -7,8 +7,11 @@ edition = "2018"
 [dependencies]
 atty = "0.2"
 failure = "0.1"
+futures = "0.3"
```
[the remark this answers is not in the document]
Comment (vipeller) [reply mqtt/mqttd/Cargo.toml]: removed

File: mqtt/mqtt-broker/Cargo.toml [hunk 8]
```diff
@@ -9,16 +9,20 @@ edition = "2018"
 [dependencies]
 async-trait = "0.1"
 bincode = "1.2"
+bytes = "0.5.4"
 chrono = "0.4"
 fail = "0.3"
-failure = "0.1"
 flate2 = "1.0"
```
[the remark this answers is not in the document]
Comment (vipeller) [reply mqtt/mqtt-broker/Cargo.toml]: sorted

File: mqtt/mqttd/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqttd/src/main.rs]: Let's make it separate PBI maybe?

File: mqtt/mqtt-broker/src/server.rs [hunk 9]
```diff
@@ -26,14 +26,15 @@ impl Server {
...
         self,
-        transports: Vec<TransportBuilder<A>>,
+        transports: I,
         shutdown_signal: F,
     ) -> Result<BrokerState, Error>
     where
-        F: Future<Output = ()> + Unpin,
         A: ToSocketAddrs,
+        F: Future<Output = ()> + Unpin,
+        I: IntoIterator<Item=TransportBuilder<A>>,
```
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqtt-broker/src/server.rs]: and after that `cargo clippy --all --test --examples`

File: mqtt/mqtt-broker/src/server.rs [hunk 10]
```diff
@@ -66,7 +67,9 @@ impl Server {
 
                 debug!("sending stop signal for every protocol head");
                 for itx in incoming_tasks_tx {
-                    itx.send(()).unwrap();
+                    if let Err(()) = itx.send(()) {
```
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqtt-broker/src/server.rs]: to reduce code repeatition you may write a macros `try_stop_protocol_head!(itx)`

File: mqtt/mqttd/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqttd/src/main.rs]: iotedged has `-c` flag instead. we probably should be consistent

File: mqtt/mqtt-broker/src/server.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (vipeller) [reply mqtt/mqtt-broker/src/server.rs]: send() doesn't seem to return more than the unit type in case of error

[the remark this answers is not in the document]
Comment (vipeller) [reply mqtt/mqtt-broker/src/server.rs]: moved the loop into a function

File: mqtt/mqtt-broker/src/server.rs [hunk 11]
```diff
@@ -28,22 +26,35 @@ impl Server {
...
+        A: ToSocketAddrs,
         F: Future<Output = ()> + Unpin,
     {
         let Server { broker } = self;
         let mut handle = broker.handle();
+        let broker_task = tokio::spawn(broker.run());
 
-        let (itx, irx) = oneshot::channel::<()>();
+        let mut incoming_tasks = Vec::new();
+        let mut incoming_tasks_tx = Vec::new();
```
[the remark this answers is not in the document]
Comment (vipeller) [reply mqtt/mqtt-broker/src/server.rs]: renamed

File: mqtt/mqttd/src/main.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (vipeller) [reply mqtt/mqttd/src/main.rs]: ok, I modified it to read sync and not async

File: mqtt/mqtt-broker/src/server.rs [hunk 12]
```diff
@@ -56,36 +68,85 @@ where
...
+                    let mut results = vec![result];
+                    results.extend(future::join_all(unfinished_incoming_tasks).await);
+
                     handle.send(Message::System(SystemEvent::Shutdown)).await?;
-                    broker_task.await.context(ErrorKind::TaskJoin)?;
-                    return Err(e);
+
+                    let broker_state = broker_task.await;
+
+                    // todo refactor to possible gather all errors
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/server.rs]: What about this TODO?

File: mqtt/mqttd/Cargo.toml [hunk 13]
```diff
@@ -8,7 +8,9 @@ edition = "2018"
 atty = "0.2"
 failure = "0.1"
 futures-util = { version = "0.3", features = ["sink"] }
-tokio = { version = "0.2", features = ["dns", "macros", "rt-threaded", "signal", "stream", "tcp", "time"] }
+tokio = { version = "0.2", features = ["dns", "macros", "rt-threaded", "signal", "stream", "tcp", "time", "fs"] }
```
Comment (dmolokanov) [inline mqtt/mqttd/Cargo.toml]: you can safely remove "fs" from features if we are not going to use async filesystem operations

File: mqtt/mqttd/Cargo.toml [hunk 14]
```diff
@@ -8,7 +8,9 @@ edition = "2018"
...
+tokio-tls = "0.3"
```
Comment (dmolokanov) [inline mqtt/mqttd/Cargo.toml]: I believe we don't need tokio-tls import for mqttd, right?

File: mqtt/mqtt-broker/src/server.rs (hunk 12 above)
Comment (vipeller) [reply mqtt/mqtt-broker/src/server.rs]: now it logs the errors and passes back the broker state as in case of the rest of the error cases

File: mqtt/mqtt-broker/src/server.rs [hunk 15]
```diff
@@ -56,36 +68,84 @@ where
...
             }
-            Either::Right((either, _shutdown)) => match either {
-                Either::Right((Ok(_incoming_task), broker_task)) => {
-                    debug!("sending Shutdown message to broker");
-                    handle.send(Message::System(SystemEvent::Shutdown)).await?;
-                    broker_task.await.context(ErrorKind::TaskJoin)?
-                }
-                Either::Right((Err(e), broker_task)) => {
-                    error!(message = "an error occurred in the accept loop", error=%e);
+            Either::Right((either, _)) => match either {
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-broker/src/server.rs]: it's really hard to follow all these nested selects... some inline comments would be nice :)
